File: RSF To Map/VBto VBNET/modRpgwoMapFormat.vb
'Option Strict Off
'Option Explicit On

Imports System.Text
Imports System.Runtime.InteropServices
Imports Microsoft.VisualBasic

Public Module modRpgwoMapFormat


	'=========================================================
#Const IndexCurFile = 2
#If CompileAll Or CompileAllBAS Or ((IndexCurFile >= CompileFilesIndexMin) And (IndexCurFile <= CompileFilesIndexMax)) Then
#Const CompileAll_modRpgwoMapFormat = True
#Else
'#Const CompileAll_modRpgwoMapFormat = True
#End If
    ' VBto upgrade warning: mMapWidth As Short	OnWrite(String)
    Public mMapWidth As Short
    ' VBto upgrade warning: mMapHeight As Short	OnWrite(String)
    Public mMapHeight As Short
    ' VBto upgrade warning: mMapVersion As StringBuilder	OnWrite(String)
    Public mMapVersion As New StringBuilder(StrDup(30, Chr(0)), 30)
    Public mMapExtra As New StringBuilder(StrDup(64, Chr(0)), 64)
    Public mSurfaceArray As Boolean
    Public mMapNotes As New StringBuilder(StrDup(5000, Chr(0)), 5000)
    Public mNumberOfItems As Short
    Public mNumberOfMonsters As Short

    Public mv2NumberOfItems As Integer
    Public mv2NumberOfMonsters As Integer


    Public v2Map As Boolean

    <StructLayout(LayoutKind.Sequential, Pack:=1)> _
    Structure mItemType
        Dim a As Boolean
        Dim c As Short
        Dim ItemID As Short
        Dim Xpos As Short
        Dim Ypos As Short
        Dim Data1 As Short
        Dim Data2 As Short
        Dim Data3 As Short
        Dim Data4 As Short
        Dim Data5 As Short
        <MarshalAs(UnmanagedType.ByValTStr, SizeConst:=14), VBFixedString(14)> Dim junk As String
        Dim Reset As Short
        Dim Uses As Short
        Dim Trigger As Short
        'Junk2 As String * 4
        Dim SpawnTimeout As Single
        ' VBto upgrade warning: Writing As StringBuilder	OnWrite(String)
        <MarshalAs(UnmanagedType.ByValTStr, SizeConst:=200), VBFixedString(200)> Dim Writing As String
        <MarshalAs(UnmanagedType.ByValTStr, SizeConst:=101), VBFixedString(101)> Dim junk3 As String
    End Structure
    <StructLayout(LayoutKind.Sequential, Pack:=1)> _
    Structure mMonsterType
        Dim a As Boolean
        Dim MonsterId As Short
        Dim Xpos As Short
        Dim Ypos As Short
        Dim Timeout As Short
        <MarshalAs(UnmanagedType.ByValTStr, SizeConst:=31), VBFixedString(31)> Dim Junk10 As String
    End Structure


    'V2/v3 RPGWO
    Structure Montype2
        Dim d As StringBuilder
        Dim MonsterName As StringBuilder
        Dim Image As Short
        Dim ImageType As Short
        Dim Level As Short
        Dim test As StringBuilder

        Public Sub New(ByVal unusedParam As Integer)
            d = New StringBuilder(StrDup(4, Chr(0)), 4)
            MonsterName = New StringBuilder(StrDup(50, Chr(0)), 50)
            test = New StringBuilder(StrDup(102, Chr(0)), 102)
        End Sub
    End Structure
    Public Monsters2() As Montype2


    Structure mMapDataType
        Dim Xpos As Short
        Dim Ypos As Short
        Dim Surface As Short
    End Structure
    Dim tempmMapData() As mMapDataType

    Public mMap(,) As Short
    Public mItems() As mItemType
    Public mMonsters() As mMonsterType



    Sub SaveRpgwoMap(ByVal filename As String)
'#Const Compile_SaveRpgwoMap = True
#If Compile_SaveRpgwoMap Or CompileAll_modRpgwoMapFormat Then
        Dim f As Integer
        f = FreeFile()
        mSurfaceArray = False

        If FileExists(filename) Then
            Kill(filename)
        End If

        FileOpen(f, filename, OpenMode.Binary)
        FilePut(f, mMapWidth)
        FilePut(f, mMapHeight)
        FilePut(f, mMapVersion.ToString(), , True)
        FilePut(f, mSurfaceArray)
        FilePut(f, mMapExtra.ToString(), , True)
        mNumberOfMonsters -= 1
        mNumberOfItems -= 1

        mv2NumberOfItems -= 1
        mv2NumberOfMonsters -= 1
        'Get Tile Id
        'ReDim mMap(1 To mMapWidth, 1 To mMapHeight)
        'Put #f, , mMap


        Dim MapSizeTotalCount As Integer
        Dim unk1 As Short
        Dim countSurfaceItems As Short
        countSurfaceItems = 1


        ReDim tempmMapData(countSurfaceItems)
        MapSizeTotalCount = 0
        Dim x As Short
        Dim y As Short
        For x = 1 To mMapWidth
            For y = 1 To mMapHeight
                If mMap(x, y)<>0 Then

                    tempmMapData(UBound(tempmMapData)).Xpos = x
                    tempmMapData(UBound(tempmMapData)).Ypos = y
                    tempmMapData(UBound(tempmMapData)).Surface = mMap(x, y)
                    MapSizeTotalCount += 1
                    ReDim Preserve tempmMapData(UBound(tempmMapData)+1)
                End If
            Next y
        Next x
        ReDim Preserve tempmMapData(UBound(tempmMapData)-1)

        If mSurfaceArray=True Then
            FilePut(f, mapElements)
            FilePut(f, tempmMapData)
        Else


            FilePut(f, mMap)
        End If


        If v2Map=True Then


            FilePut(f, mv2NumberOfItems)
            ' MsgBox mNumberOfItems
            'ReDim mItems(1 To mNumberOfItems)
            If mv2NumberOfItems>0 Then
                ReDim Preserve mItems(modRpgwoMapFormat.mv2NumberOfItems)
                FilePut(f, mItems)
            End If
            'MsgBox Loc(1)
            FilePut(f, mv2NumberOfMonsters)
            'MsgBox mNumberOfMonsters
            'ReDim mMonsters(1 To mNumberOfMonsters)
            If mv2NumberOfMonsters>0 Then
                ReDim Preserve mMonsters(modRpgwoMapFormat.mv2NumberOfMonsters)
                FilePut(f, mMonsters)
            End If


        Else

            FilePut(f, mNumberOfItems)
            ' MsgBox mNumberOfItems
            'ReDim mItems(1 To mNumberOfItems)
            If mNumberOfItems>0 Then
                ReDim Preserve mItems(modRpgwoMapFormat.mNumberOfItems)
                FilePut(f, mItems)
            End If



            FilePut(f, mNumberOfMonsters)

            'MsgBox mNumberOfMonsters
            'ReDim mMonsters(1 To mNumberOfMonsters)
            If mNumberOfMonsters>0 Then
                ReDim Preserve mMonsters(modRpgwoMapFormat.mNumberOfMonsters)
                FilePut(f, mMonsters)
            End If


        End If
        ' MsgBox "notes start:" & Loc(f)
        Dim k As Short
        k = 5000
        FilePut(f, k)
        ' VBto upgrade warning: notes As StringBuilder	OnWrite(String)
        Dim notes As New StringBuilder(StrDup(5000, Chr(0)), 5000)
        StrCpy(notes, Space(k))
        FilePut(f, notes.ToString(), , True)
        'Get Notes
        FileClose(f)
#End If	' Compile_SaveRpgwoMap
    End Sub

End Module

File: RSF To Map_.Net/Upgraded/modRpgwoMapFormat.cs
using Microsoft.VisualBasic;
using System;
using System.IO;
using System.Runtime.InteropServices;
using UpgradeHelpers.Helpers;

namespace RsfToMap
{
	internal static class modRpgwoMapFormat
	{

		public static short mMapWidth = 0;
		public static short mMapHeight = 0;
		public static string mMapVersion = new string('\0', 30);
		public static string mMapExtra = new string('\0', 64);
		public static short mSurfaceArray = 0;
		public static string mMapNotes = new string('\0', 5000);
		public static short mNumberOfItems = 0;
		public static short mNumberOfMonsters = 0;

		public static int mv2NumberOfItems = 0;
		public static int mv2NumberOfMonsters = 0;


		public static bool v2Map = false;

		[Serializable][StructLayout(LayoutKind.Sequential, CharSet=CharSet.Ansi)]
		public struct mItemType
		{
			public bool a;
			public short c;
			public short ItemID;
			public short Xpos;
			public short Ypos;
			public short Data1;
			public short Data2;
			public short Data3;
			public short Data4;
			public short Data5;
			[Microsoft.VisualBasic.VBFixedArray(13)][System.Runtime.InteropServices.MarshalAs(System.Runtime.InteropServices.UnmanagedType.ByValArray, SizeConst=14)]
			public char[] _junk;
			public short Reset;
			public short Uses;
			public short Trigger;
			//Junk2 As String * 4
			public float SpawnTimeout;
			[Microsoft.VisualBasic.VBFixedArray(199)][System.Runtime.InteropServices.MarshalAs(System.Runtime.InteropServices.UnmanagedType.ByValArray, SizeConst=200)]
			public char[] _Writing;
			[Microsoft.VisualBasic.VBFixedArray(100)][System.Runtime.InteropServices.MarshalAs(System.Runtime.InteropServices.UnmanagedType.ByValArray, SizeConst=101)]
			public char[] _junk3;
			public string junk
			{
				get
				{
					return new string(_junk);
				}
				set
				{
					ArraysHelper.CopyStringToCharArray(_junk, value);
				}
			}

			public string Writing
			{
				get
				{
					return new string(_Writing);
				}
				set
				{
					ArraysHelper.CopyStringToCharArray(_Writing, value);
				}
			}

			public string junk3
			{
				get
				{
					return new string(_junk3);
				}
				set
				{
					ArraysHelper.CopyStringToCharArray(_junk3, value);
				}
			}

			public static mItemType CreateInstance()
			{
				mItemType result = new mItemType();
				result._junk = new char[14];
				result._Writing = new char[200];
				result._junk3 = new char[101];
				return result;
			}
		}
		[Serializable][StructLayout(LayoutKind.Sequential, CharSet=CharSet.Ansi)]
		public struct mMonsterType
		{
			public bool a;
			public short MonsterId;
			public short Xpos;
			public short Ypos;
			public short Timeout;
			[Microsoft.VisualBasic.VBFixedArray(30)][System.Runtime.InteropServices.MarshalAs(System.Runtime.InteropServices.UnmanagedType.ByValArray, SizeConst=31)]
			public char[] _Junk10;
			public string Junk10
			{
				get
				{
					return new string(_Junk10);
				}
				set
				{
					ArraysHelper.CopyStringToCharArray(_Junk10, value);
				}
			}

			public static mMonsterType CreateInstance()
			{
				mMonsterType result = new mMonsterType();
				result._Junk10 = new char[31];
				return result;
			}
		}


		//V2/v3 RPGWO
		[Serializable][StructLayout(LayoutKind.Sequential, CharSet=CharSet.Ansi)]
		public struct Montype2
		{
			[Microsoft.VisualBasic.VBFixedArray(3)][System.Runtime.InteropServices.MarshalAs(System.Runtime.InteropServices.UnmanagedType.ByValArray, SizeConst=4)]
			public char[] _d;
			[Microsoft.VisualBasic.VBFixedArray(49)][System.Runtime.InteropServices.MarshalAs(System.Runtime.InteropServices.UnmanagedType.ByValArray, SizeConst=50)]
			public char[] _MonsterName;
			public short Image;
			public short ImageType;
			public short Level;
			[Microsoft.VisualBasic.VBFixedArray(101)][System.Runtime.InteropServices.MarshalAs(System.Runtime.InteropServices.UnmanagedType.ByValArray, SizeConst=102)]
			public char[] _test;
			public string d
			{
				get
				{
					return new string(_d);
				}
				set
				{
					ArraysHelper.CopyStringToCharArray(_d, value);
				}
			}

			public string MonsterName
			{
				get
				{
					return new string(_MonsterName);
				}
				set
				{
					ArraysHelper.CopyStringToCharArray(_MonsterName, value);
				}
			}

			public string test
			{
				get
				{
					return new string(_test);
				}
				set
				{
					ArraysHelper.CopyStringToCharArray(_test, value);
				}
			}

			public static Montype2 CreateInstance()
			{
				Montype2 result = new Montype2();
				result._d = new char[4];
				result._MonsterName = new char[50];
				result._test = new char[102];
				return result;
			}
		}
		public static Montype2[] Monsters2 = null;


		[Serializable]
		private struct mMapDataType
		{
			public short Xpos;
			public short Ypos;
			public short Surface;
		}
		static mMapDataType[] tempmMapData = null;

		public static short[, ] mMap = null;
		public static mItemType[] mItems = null;
		public static mMonsterType[] mMonsters = null;



		internal static void SaveRpgwoMap(string filename)
		{
			object mapElements = null;
			int f = FileSystem.FreeFile();
			mSurfaceArray = 0;

			if (modGlobals.FileExists(filename))
			{
				File.Delete(filename);
			}

			FileSystem.FileOpen(f, filename, OpenMode.Binary, OpenAccess.Write, OpenShare.LockWrite, -1);
			FileSystem.FilePut(f, mMapWidth);
			FileSystem.FilePut(f, mMapHeight);
			FileSystem.FilePut(f, mMapVersion);
			FileSystem.FilePut(f, mSurfaceArray);
			FileSystem.FilePut(f, mMapExtra);
			mNumberOfMonsters--;
			mNumberOfItems--;

			mv2NumberOfItems--;
			mv2NumberOfMonsters--;
			//Get Tile Id
			//ReDim mMap(1 To mMapWidth, 1 To mMapHeight)
			//Put #f, , mMap


			int countSurfaceItems = 1;


			tempmMapData = ArraysHelper.InitializeArray<mMapDataType>(countSurfaceItems);
			int MapSizeTotalCount = 0;
			int tempForEndVar = mMapWidth;
			for (int x = 1; x <= tempForEndVar; x++)
			{
				int tempForEndVar2 = mMapHeight;
				for (int y = 1; y <= tempForEndVar2; y++)
				{
					if (mMap[x - 1, y - 1] != 0)
					{

						tempmMapData[tempmMapData.GetUpperBound(0)].Xpos = (short) x;
						tempmMapData[tempmMapData.GetUpperBound(0)].Ypos = (short) y;
						tempmMapData[tempmMapData.GetUpperBound(0)].Surface = (short) mMap[x - 1, y - 1];
						MapSizeTotalCount++;
						tempmMapData = ArraysHelper.RedimPreserve(tempmMapData, new int[]{tempmMapData.GetUpperBound(0) + 2});
					}
				}
			}
			tempmMapData = ArraysHelper.RedimPreserve(tempmMapData, new int[]{tempmMapData.GetUpperBound(0)});

			if (mSurfaceArray == 1)
			{
				FileSystem.FilePut(f, mapElements);
				FileSystem.FilePut(f, tempmMapData);
			}
			else
			{


				FileSystem.FilePut(f, mMap);
			}


			if (v2Map)
			{


				FileSystem.FilePut(f, mv2NumberOfItems);
				// MsgBox mNumberOfItems
				//ReDim mItems(1 To mNumberOfItems)
				if (mv2NumberOfItems > 0)
				{
					mItems = ArraysHelper.RedimPreserve(mItems, new int[]{modRpgwoMapFormat.mv2NumberOfItems});
					FileSystem.FilePut(f, mItems);
				}
				//MsgBox Loc(1)
				FileSystem.FilePut(f, mv2NumberOfMonsters);
				//MsgBox mNumberOfMonsters
				//ReDim mMonsters(1 To mNumberOfMonsters)
				if (mv2NumberOfMonsters > 0)
				{
					mMonsters = ArraysHelper.RedimPreserve(mMonsters, new int[]{modRpgwoMapFormat.mv2NumberOfMonsters});
					FileSystem.FilePut(f, mMonsters);
				}


			}
			else
			{

				FileSystem.FilePut(f, mNumberOfItems);
				// MsgBox mNumberOfItems
				//ReDim mItems(1 To mNumberOfItems)
				if (mNumberOfItems > 0)
				{
					mItems = ArraysHelper.RedimPreserve(mItems, new int[]{modRpgwoMapFormat.mNumberOfItems});
					FileSystem.FilePut(f, mItems);
				}



				FileSystem.FilePut(f, mNumberOfMonsters);

				//MsgBox mNumberOfMonsters
				//ReDim mMonsters(1 To mNumberOfMonsters)
				if (mNumberOfMonsters > 0)
				{
					mMonsters = ArraysHelper.RedimPreserve(mMonsters, new int[]{modRpgwoMapFormat.mNumberOfMonsters});
					FileSystem.FilePut(f, mMonsters);
				}


			}
			// MsgBox "notes start:" & Loc(f)
			short k = 5000;
			FileSystem.FilePut(f, k);
			string notes = new string('\0', 5000);
			notes = StringsHelper.GetFixedLengthString(new String(' ', k), 5000);
			FileSystem.FilePut(f, notes);
			//Get Notes
			FileSystem.FileClose(f);
		}
	}
}